pkg/client/didexchange/client.go

```diff
@@ -119,35 +119,134 @@ func (c *Client) RemoveConnection(id string) error {
 }
 
 // startServiceEventListener listens to action and message events from DID Exchange service.
-func (c *Client) startServiceEventListener() error {
-	// register the action event channel
+func (c *Client) startServiceEventListener() {
+	// listen for action event and message events
+	go func() {
+		for {
+			select {
+			case e := <-c.actionCh:
+				// assigned to var as lint fails with : Using a reference for the variable on range scope (scopelint)
+				msg := e
+				c.handleActionEvent(&msg)
+			case e := <-c.msgCh:
+				// assigned to var as lint fails with : Using a reference for the variable on range scope (scopelint)
+				msg := e
+				c.handleMessageEvent(&msg)
+			}
+		}
+	}()
+}
+
+// RegisterActionEvent on DID Exchange protocol messages. The events are triggered for incoming exchangeRequest,
+// exchangeResponse and exchangeAck message types. The consumer need to invoke the callback to resume processing.
+// Only one channel can be registered for the action events. The function will throw error if a channel is already
+// registered. The AutoExecuteActionEvent() function can be used to automatically trigger callback function for the
+// event.
+func (c *Client) RegisterActionEvent(ch chan<- dispatcher.DIDCommAction) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if c.actionEvent != nil {
+		return errors.New("channel is already registered for the action event")
+	}
+
+	c.actionEvent = ch
+
+	// register the service action event channel
 	err := c.didexchangeSvc.RegisterActionEvent(c.actionCh)
 	if err != nil {
+		c.actionEvent = nil
 		return fmt.Errorf("didexchange action event registration failed: %w", err)
 	}
 
-	// register the message event channel
-	err = c.didexchangeSvc.RegisterMsgEvent(c.msgCh)
+	return nil
+}
+
+// UnregisterActionEvent on DID Exchange protocol messages. Refer RegisterActionEvent().
+func (c *Client) UnregisterActionEvent(ch chan<- dispatcher.DIDCommAction) error {
+	c.lock.Lock()
+	defer c.lock.Unlock()
+
+	if c.actionEvent != ch {
+		return errors.New("invalid channel passed to unregister the action event")
+	}
+
+	c.actionEvent = nil
+
+	// unregister the service action event channel
+	err := c.didexchangeSvc.UnregisterActionEvent(c.actionCh)
+	if err != nil {
+		c.actionEvent = ch
+		return fmt.Errorf("didexchange action event unregistration failed: %w", err)
+	}
+
+	return nil
+}
+
+// RegisterMsgEvent on DID Exchange protocol messages. The message events are triggered for state transitions. Client
+// will not expect any callback on these events unlike Action events.
+func (c *Client) RegisterMsgEvent(ch chan<- dispatcher.StateMsg) error {
+	// register the service message event channel
+	err := c.didexchangeSvc.RegisterMsgEvent(c.msgCh)
```

Comment:
If this is called multiple times then you'll register the same channel (c.msgCh) multiple times.

Reply:
Valid point. [Added](https://github.com/hyperledger/aries-framework-go/compare/d58b6228a75eddb0dd13d89c07ec443bf5b490ec..7f2cd61c8bc3526f13b02f38f39e9946c1881f10#diff-427522b353a28b1ec350c29c79d387b0R192) a condition to register only once.